src/widget/wtracktableview.cpp

```diff
@@ -262,28 +258,31 @@ void WTrackTableView::loadTrackModel(QAbstractItemModel* model) {
                 &WTrackTableView::slotSortingChanged,
                 Qt::AutoConnection);
 
-        int sortColumn;
         Qt::SortOrder sortOrder;
-
-        // Stupid hack that assumes column 0 is never visible, but this is a weak
-        // proxy for "there was a saved column sort order"
-        if (horizontalHeader()->sortIndicatorSection() > 0) {
+        TrackModel::SortColumnId sortColumn =
+                trackModel->sortColumnIdFromColumnIndex(
+                        horizontalHeader()->sortIndicatorSection());
+        if (sortColumn != TrackModel::SortColumnId::Invalid) {
             // Sort by the saved sort section and order.
-            sortColumn = horizontalHeader()->sortIndicatorSection();
             sortOrder = horizontalHeader()->sortIndicatorOrder();
         } else {
             // No saved order is present. Use the TrackModel's default sort order.
-            sortColumn = trackModel->defaultSortColumn();
+            sortColumn = trackModel->sortColumnIdFromColumnIndex(trackModel->defaultSortColumn());
             sortOrder = trackModel->defaultSortOrder();
 
             // If the TrackModel has an invalid or internal column as its default
-            // sort, find the first non-internal column and sort by that.
-            while (sortColumn < 0 || trackModel->isColumnInternal(sortColumn)) {
-                sortColumn++;
+            // sort, find the first valid sort column and sort by that.
+            int sortColumnIndex = -1;
+            constexpr int kMaxProbeIndex = 10; // just to avoid an endless while loop
+            while (sortColumn == TrackModel::SortColumnId::Invalid &&
+                    sortColumnIndex < kMaxProbeIndex) {
+                sortColumnIndex++;
+                sortColumn = trackModel->sortColumnIdFromColumnIndex(sortColumnIndex);
             }
+            DEBUG_ASSERT(sortColumn != TrackModel::SortColumnId::Invalid);
```

Comment:
Using a debug assert is wrong, because it will trigger when reaching kMaxProbeIndex, which is possible.

Why don't you need an artificial constant instead of using model()->columnCount() as the upper limit?